### PR TITLE
Enhanced Target Server Validator

### PR DESCRIPTION
**Pull request by payaljindal**
## Description
What's changed, or what was fixed?

- Updated the target proxy payload from one host_port combination to a json payload with multiple host_port combinations using Gson library
- Added JSON Threat Protection Policy so that the JSON payload doesn't have more than 10 host_port combinations
- Added run_parallel function for fetching target servers, fetching proxy bundles and making calls to the newly created target server proxy
- Updated prints to logger
- Fixed Target Apigee Type

## Issues Fixed
- N/A

## Housekeeping
(please check all that apply [x], do not edit the text)
- [ ] I have run all the tests locally and they all pass.
- [ ] I have followed the relevant style guide for my changes.

## Full Repo Validation Required
(please check all that apply [x], do not edit the text)
- [ ] PR requires full pipeline run (Run for changes only by default).

**CC:** @apigee-devrel-reviewers

### CODE REVIEW

File: tools/target-server-validator/apigee_utils.py [hunk 1]
```diff
@@ -154,10 +157,10 @@ def get_api_revisions_deployment(self, env, api_name, api_rev):  # noqa
             if self.apigee_type == "opdk":
                 if api_deployment_status == "deployed":
                     return True
-            print(f"API {api_name} is in Status: {api_deployment_status} !")  # noqa
+            logger.debug(f"API {api_name} is in Status: {api_deployment_status} !")  # noqa
             return False
         else:
-            print(response.text)
+            logger.debug(f"{response.text}")
```
Comment (danistrebel) [inline tools/target-server-validator/apigee_utils.py:163]: Does this need an f-string?

File: tools/target-server-validator/callout/src/main/java/com/google/apigee/json/JavaxJson.java [hunk 2]
```diff
@@ -0,0 +1,190 @@
+// JavaxJson.java
```
Comment (danistrebel) [inline tools/target-server-validator/callout/src/main/java/com/google/apigee/json/JavaxJson.java:1]: License Header needed here

File: tools/target-server-validator/callout/src/main/java/com/google/apigee/json/JavaxJson.java [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+import javax.json.JsonValue.ValueType;
+
+public class JavaxJson {
+  @SuppressWarnings("unchecked")
+  public static <T> T fromJson(String json, Class<T> beanClass) {
+    JsonValue value = Json.createReader(new StringReader(json)).read();
+    return (T) decode(value, beanClass);
+  }
+
+  private static Object decode(JsonValue jsonValue, Type targetType) {
```
Comment (danistrebel) [inline tools/target-server-validator/callout/src/main/java/com/google/apigee/json/JavaxJson.java:34]: What's the rationale for mixing getValueType Enums and instanceof? Would it be more readable to just use the enum and turn it into a switch?

File: tools/target-server-validator/callout/src/main/java/PortOpenCheck.java [hunk 4]
```diff
@@ -72,12 +95,34 @@ private static String available(final String host, final int port) {
   public ExecutionResult execute(final MessageContext messageContext,
     final ExecutionContext executionContext) {
     try {
-      String hostname = messageContext.getMessage().getHeader("host_name");
-      String port = messageContext.getMessage().getHeader("port_number");
-      int portnumber = Integer.parseInt(port);
-      String status = available(hostname, portnumber);
-      messageContext.setVariable("flow.reachableStatus", status);
-      return ExecutionResult.SUCCESS;
+      String payload = (String) messageContext.getVariable("request.content");
```
Comment (danistrebel) [inline tools/target-server-validator/callout/src/main/java/PortOpenCheck.java:98]: Quite a complex datastructure in the request. Can we add some docs on the stucture of the payload?

File: tools/target-server-validator/callout/src/main/java/PortOpenCheck.java [hunk 5]
```diff
@@ -72,12 +95,34 @@ private static String available(final String host, final int port) {
...
+      if (payload != null) {
+        Map<String, List<Map<String, String>>> result = new HashMap<>();
+        result.put("hostname_portnumbers_status", new ArrayList<>());
+        Map<String,List<Map<String,String>>> outerMap  = JavaxJson.fromJson(payload,Map.class);
```
Comment (danistrebel) [inline tools/target-server-validator/callout/src/main/java/PortOpenCheck.java:102]: What's the motivation for implementing a generic json parser if we know the json format to be expected. Can't  we traversing the known structure with a JsonReader.readMap() and continue from there the same way as you do with the  explicit mapping operations below?

**Review by danistrebel — CHANGES_REQUESTED** [review]

File: tools/target-server-validator/callout/src/main/java/PortOpenCheck.java (hunk 4 above)
Comment (OmidTahouri) [reply tools/target-server-validator/callout/src/main/java/PortOpenCheck.java:98]: +1 please update the readme to include example JSON for both the request and response

File: tools/target-server-validator/callout/src/main/java/PortOpenCheck.java (hunk 5 above)
Comment (OmidTahouri) [reply tools/target-server-validator/callout/src/main/java/PortOpenCheck.java:102]: +1 Can we not simply parse the entire JSON into HashMaps and ArrayLists with a single operation? I think this approach is suitable if unmarshalling into custom classes, but gson (and other libraries) should natively parse into HashMaps and ArrayLists.

**Review by OmidTahouri — CHANGES_REQUESTED** [review]

File: tools/target-server-validator/apigee_utils.py (hunk 1 above)
Comment (payaljindal) [reply tools/target-server-validator/apigee_utils.py]: Changed it.

File: tools/target-server-validator/callout/src/main/java/PortOpenCheck.java (hunk 4 above)
Comment (payaljindal) [reply tools/target-server-validator/callout/src/main/java/PortOpenCheck.java:78]: Updated the README.

File: tools/target-server-validator/callout/src/main/java/PortOpenCheck.java (hunk 5 above)
Comment (payaljindal) [reply tools/target-server-validator/callout/src/main/java/PortOpenCheck.java]: Right, I have updated the library to Gson and removed the generic parsing.

**Review by danistrebel — APPROVED** [review]

**Review by OmidTahouri — APPROVED** [review]

**Pull request merged**